Fix #21889: Blog page button in teach-page now leading to the right domain

daniel-pintoo opened the pull request:
"www.oppia.org" being hardcoded caused the issue, now the domain is being inherited from the current page.

Files changed fixing the bug:
-app.constants.ts:  removed 'TEACHER_STORY_TAGGED_BLOGS_LINK' since
    it is not being used anymore
-teach-page.component.ts:  updated 'teacherStoryTaggedBlogsLink' to
    use a relative domain

Files changed for testing the fix:
-teach-page.component.spec.ts:  new test checking if relative URL is
    being used
-click-all-buttons-on-teach-page.spec.ts:  new test clicking
    'Check out our blog' button
-logged-out-user.ts:  switched 'clickBlogButtonInTeachPage()' from
    opening an external link to waiting for a click
    on '.e2e-test-teach-page-blog-button'

## Overview

1. This PR fixes #21889.
2. This PR does the following: Changes the blog link in the teach-page so that is has a dynamic domain.
3. The original bug occurred because: the "www.oppia.org" domain was hardcoded into the link.

## Essential Checklist

- [x] I have linked the issue that this PR fixes in the "Development" section of the sidebar.
- [x] I have checked the "Files Changed" tab and confirmed that the changes are what I want to make.
- [x] I have written tests for my code.
- [x] The **PR title** starts with "Fix #bugnum: " or "Fix part of #bugnum: ...", followed by a short, clear summary of the changes.
- [ ] I have assigned the correct reviewers to this PR (or will leave a comment with the phrase "@{{reviewer_username}} PTAL" if I can't assign them directly).

## Proof that changes are correct

Before video:

https://github.com/user-attachments/assets/aa947b4a-596c-4e83-9142-63164cd1401a

After video:

https://github.com/user-attachments/assets/1cb9b853-d02a-4c31-9115-4135db5735b5

#### Proof of changes on desktop with slow/throttled network

#### Proof of changes on mobile phone

#### Proof of changes in Arabic language

## PR Pointers

daniel-pintoo committed:
Blog page button in teach-page now leading to the right domain
"www.oppia.org" being hardcoded caused the issue, now the domain
is being inherited from the current page.

Files changed fixing the bug:
-app.constants.ts:  removed 'TEACHER_STORY_TAGGED_BLOGS_LINK' since
    it is not being used anymore
-teach-page.component.ts:  updated 'teacherStoryTaggedBlogsLink' to
    use a relative domain

Files changed for testing the fix:
-teach-page.component.spec.ts:  new test checking if relative URL is
    being used
-click-all-buttons-on-teach-page.spec.ts:  new test clicking
    'Check out our blog' button
-logged-out-user.ts:  switched 'clickBlogButtonInTeachPage()' from
    opening an external link to waiting for a click
    on '.e2e-test-teach-page-blog-button'

Nik-09 reviewed (approved):
Thank you @daniel-pintoo LGTM

imchristie commented on core/tests/puppeteer-acceptance-tests/utilities/user/logged-out-user.ts:
```diff
@@ -1495,10 +1492,11 @@ export class LoggedOutUser extends BaseUser {
-      blogButtonInTeachPage,
-      teacherStoryTaggedBlogsLink
-    );
+    await this.clickOn('.e2e-test-teach-page-blog-button');
```
Store this html element as const.

imchristie commented on core/tests/puppeteer-acceptance-tests/utilities/user/logged-out-user.ts:
```diff
@@ -1495,10 +1492,11 @@ export class LoggedOutUser extends BaseUser {
+    await this.clickOn('.e2e-test-teach-page-blog-button');
+  }
+
+  async getCurrentUrl(): Promise<string> {
```
I think it's better to put this function in puppeteer-utils.ts since this function is not fixed for logged-out-user only.

imchristie replied:
You still haven't remove this function from this file logged-out-user.ts.

imchristie reviewed (changes requested):
Left some comments. The CI checks are failing. Please look into that as well.

daniel-pintoo commented:
Hello, I am having trouble with a problem not related with the changes requested. I made the changes without any issue, but when trying to push them into my fork, I ran into a problem while running the tests in the pre-push hook, which stopped the push.

Here is a screenshot of the error:

![Screenshot from 2025-04-01 15-39-19](https://github.com/user-attachments/assets/3f94ccc1-2100-4ee2-950f-8175e4a25d49)

The problem is also occuring when i try to run the server locally.

From what I have researched, the problem seems to be related to a TypeScript version incompatibility, but my virtual environment is running with JavaScript 4.1.6, which is the same it has always been and should run the server without any problems.

I have been trying to solve the problem without success, these are some fixes i tried:
-undo the changes (figured they caused it, but could not even run with the exact same version of the code from the last commit);
-cloning the project in a new folder with a new virtual environment (resulted in the same error);
-updating the local typescript version (could not do it due to dependency issues);
-manually changing the lines that caused trouble (the file is autogenerated every time so the changes are written over);

To whomever sees this message, i would like to ask if anyone has ever mentioned this issue and, if not, to ask for assistance on how to proceed: either to fix this error on my machine, or to find a way for going arround the error when pushing the changes to my fork.

I appreciate your attention and apologize for the delay in addressing the problem.

PS: I have an open issue on this matter in the Oppia Q&A (Developer Workflow), waiting for a response.

daniel-pintoo committed:
Requested changes implemented (PR #22245)
'.e2e-test-teach-page-blog-button' stored as constant in logged-out-user.ts
"getCurrentURL()" moved to puppeteer-utils.ts
added:
"resolutions": {
    "@types/babel__traverse": "7.20.6"
  }
to the end of package.json to fix TypeScript compatibility issues

daniel-pintoo commented:
Hi @imchristie, I made the changes you requested and managed to solve the problem I mentioned in the last message.

As you can see, one of the tests is failing and two are being skipped. 

The failing one is: "Questions List Component should not disable save button when user cancels the save modal", which is not impacted by the changes made, so I assume it was already an issue before I created this PR.

In regards to the two skipped tests, I don't quite understand why it happened and would appreciate your guidance if there is something I can do about it.

daniel-pintoo commented:
Hi again @imchristie. Just checking if there’s any update or feedback on the latest changes. I appreciate your time and assistance on this.

imchristie reviewed (commented):
Left one comment.

imchristie commented:
> Hi @imchristie, I made the changes you requested and managed to solve the problem I mentioned in the last message.
> 
> As you can see, one of the tests is failing and two are being skipped.
> 
> The failing one is: "Questions List Component should not disable save button when user cancels the save modal", which is not impacted by the changes made, so I assume it was already an issue before I created this PR.
> 
> In regards to the two skipped tests, I don't quite understand why it happened and would appreciate your guidance if there is something I can do about it.

Does it still fail now? I don't see it in the CI checks now.

daniel-pintoo committed:
getCurrentUrl() removed from logged-out-user.ts

daniel-pintoo commented:
@imchristie My apologies, I undid some previous changes when trying to fix the problem I mentioned before, and accidentally left the getCurrentUrl function in logged-out-user.ts.

Already fixed it and all the tests are running without any issue.

Is there anything else I should do, or is the bug fix ready to be merged?

Thank you for your help!

seanlip commented:
@daniel-pintoo If you want reviewers to look at the PR, please ensure that the assignees list is updated and shows who is supposed to "take the next action" on the PR. Reviewers look at github.com/pulls/assigned to see what they need to review, so if they are not assigned to your PR then they will not see it.

See checkbox 5 of the "Essential checklist" in the PR description for info on how to do the above. Thanks.

imchristie reviewed (approved):
LGTM.

seanlip commented on core/tests/puppeteer-acceptance-tests/specs/logged-out-user/click-all-buttons-on-teach-page.spec.ts:
```diff
@@ -55,6 +55,19 @@ describe('Logged-out User in Teach page', function () {
+      expect(currentUrl).toContain(
+        '/blog/search/find?q=&tags=(%22Teacher%20story%22)'
+      );
+      expect(currentUrl).not.toContain('oppia.org');
```
I think it's fine to not include this assertion. It's weird to do it for this page and not all the other pages that are internal website links.

I think you can just drop this line.

seanlip reviewed (commented):
@daniel-pintoo One small comment. Otherwise LGTM, thanks!

daniel-pintoo committed:
"expect(currentUrl).not.toContain('oppia.org');" removed
from click-all-buttons-on-teach-page.spec.ts

seanlip commented:
@daniel-pintoo Please reassign reviewers when you would like them to do a review; see instructions in the "Essential checklist" at the top for doing this.

To handle CI issues, follow [these instructions](https://github.com/oppia/oppia/wiki/If-CI-checks-fail-on-your-PR).

Thanks!

daniel-pintoo commented:
> @daniel-pintoo Please reassign reviewers when you would like them to do a review; see instructions in the "Essential checklist" at the top for doing this.
> 
> To handle CI issues, follow [these instructions](https://github.com/oppia/oppia/wiki/If-CI-checks-fail-on-your-PR).
> 
> Thanks!

Will do... I'm passing all the tests on my server, but one is failing here. Didn't want to ask for assistance on this but I'm not figuring it out on my own so might aswell ask for help.

Thanks

daniel-pintoo commented:
Hello @imchristie, as you can see in the screenshot below, the failint test is related to the creators carroussel in which I never changed anything.

![Screenshot from 2025-05-08 18-23-20](https://github.com/user-attachments/assets/5296c46f-8b2f-4a25-84ac-6a950e90632f)

The last and only change I made that seems to have caused this was removing the line 
`expect(currentUrl).not.toContain('oppia.org');` from the click-all-buttons-on-teach-page.spec.ts file (my last commit).

I would appreciate some guidance in order to solve this issue.

Thanks!

seanlip commented:
@daniel-pintoo Which parts of [these instructions](https://github.com/oppia/oppia/wiki/If-CI-checks-fail-on-your-PR) have you followed?

seanlip committed:
Merge branch 'develop' into fix-blog-links

daniel-pintoo commented:
Hello @seanlip , I apologize for the late response.
I found the problem that was causing the test to fail and fixed it.
All tests ar now passing on my PC instead of the test mentioned on #20802, which is preventing me from pushing the changes to my fork.
Is there any way to work arround this?

seanlip commented:
@daniel-pintoo That's great! Sorry, I'm not sure I have an alternative to #20802 for you ... the only way forward might be to fix it. How come it only started happening now, though?

daniel-pintoo commented:
@seanlip the only possibility I can think of is that there was some backend change in the develop branch when you merged it into the my "fix-blog-links" branch, which might have caused the bug to appear.
I have a lot on my plate at the moment, so it will probably take a while before i can look at #20802.
Isn't there anyway to skip the pre-push hook, more specifically the backend tests, since my changes were only related to frontend?
